From dbc1e36173f8cac74ef276a88ba1e0bc9964002c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?St=C3=A9phane=20Graber?= <stgraber@stgraber.org>
Date: Fri, 24 Jul 2026 11:37:25 -0400
Subject: [PATCH] incusd/project: Restrict volume creation options in
 restricted projects
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

A user in a restricted project could set block.create_options to inject
arguments into the filesystem creation command run as root. Restrict such
projects to the pool's configured default for that option.

This addresses CVE-2026-62867

Signed-off-by: Stéphane Graber <stgraber@stgraber.org>
---
 internal/server/project/permissions.go | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/internal/server/project/permissions.go b/internal/server/project/permissions.go
index 4bf91dedaf6..900f3c974f4 100644
--- a/internal/server/project/permissions.go
+++ b/internal/server/project/permissions.go
@@ -318,6 +318,19 @@ func AllowVolumeCreation(tx *db.ClusterTx, projectName string, poolName string,
 		return errors.New("Restricted projects aren't allowed to use pull mode migration")
 	}
 
+	// Restricted projects can't override low-level volume options that are passed to
+	// filesystem tooling running as root; they may only use the pool's configured default.
+	if util.IsTrue(info.Project.Config["restricted"]) {
+		_, pool, _, err := tx.GetStoragePool(context.Background(), poolName)
+		if err != nil {
+			return err
+		}
+
+		if req.Config["block.create_options"] != "" && req.Config["block.create_options"] != pool.Config["volume.block.create_options"] {
+			return errors.New(`Storage volume option "block.create_options" cannot be set in a restricted project`)
+		}
+	}
+
 	// Add the volume being created.
 	info.Volumes = append(info.Volumes, db.StorageVolumeArgs{
 		Name:     req.Name,
